### PR TITLE
Farming Mode Statistics

### PR DESCRIPTION
**Pull request by Mulchman**
I know it needs localization stuff added, but wanted to start getting more feedback on this. Original thread is here: https://github.com/DestinyItemManager/DIM/issues/1189

### CODE REVIEW

**Comment by Mulchman** [comment]
Couple of thoughts-

`dimFarmItem.directive.js` is probably not needed (unless there are going to be bigger visual display changes made to the items. As it is now, `dimStoreItem.directive.js` would suffice with the single `|| vm.item.farmReport` addition so that engrams show their amount #.

After playing with it tracking, I think rep increase indicators would be really cool to add... I'm thinking something along the lines of how I'm showing glimmer/marks: just a positive number with the faction icon next to it.

**Comment by Mulchman** [comment]
There's also a bug in that there are [at least?] 2 hashes for motes of light and the UI doesn't consolidate them into one count.

**Comment by delphiactual** [comment]
This is great, especially taking into account for i18n from the start.

**Comment by Mulchman** [comment]
Idea for adding faction rep-

screen shot 2016-11-26 at 12 39 44 pm

The reputation increase values are faked for the picture, and you'd never have this many showing at one time.

**Comment by delphiactual** [comment]
@Mulchman I think you should add padding for the # of legendary marks so it doesn't shift around on the summary.

**Comment by Mulchman** [comment]
@delphiactual can you elaborate a little more on the padding? CSS padding? Pad the number with zeros so it's always triple digits? Something else?

**Comment by delphiactual** [comment]
Just enough blank space so there is room for triple digits without showing preceding zeroes.

**Comment by Mulchman** [comment]
@delphiactual ah! Gotcha.

**Comment by Mulchman** [comment]
What about just reusing `dimStoreReputation.directive.js` for faction rep?

screen shot 2016-11-26 at 8 44 19 pm

And also in this picture I have a crude background being applied indicating I ranked up Vanguard. (Need to look into the duplicate strange coin thing; not sure why it's doing that.)

**Comment by SunburnedGoose** [comment]
What's the card icon for that is in the upper-right between items and stop?

I do like the indicator that the player has ranked up and needs to take action on it.  Rather than show progress on factions all the time, just have it show up when you rank up (unless it is Iron Banner).

**Comment by delphiactual** [comment]
it toggles back and forth between the summary and the standard farming mode display

**Comment by SunburnedGoose** [comment]
Does it have a badge that let's the user know when new items/faction rank are collected while in the summary screen?  The engram dialog just shows new engrams, but this screen is a bit broader and shows the player's loot generation during activity.  Having a badge would be inviting for the user to learn about the features.

**Comment by Mulchman** [comment]
Woops - should have checked the post before committing it seems.

When on the normal/original farm screen, what would you like to show/happen @SunburnedGoose ? Should the report button glow or something to indicate activity on the summary screen? Should there be something new to denote a faction rank-up? And why would Iron Banner rep be different from other rep?

**Comment by SunburnedGoose** [comment]
Something like a bootstrap badge: A small number bubble next to the summary icon. Question is, what's useful knowledge?  We already show the # of engrams in the summary screen.  What do we want users to know about in the gather screen?  IMHO, only the faction rank ups that require a trip to the tower should be indicated by a badge next to the icon.  Everything else is just 'nice to know'.

**Comment by Mulchman** [comment]
I googled 'Bootstrap badge' and have an idea of what you mean now. If the badge is only showing rank-ups, what if the badge was the actual faction icon(s)? Still working on where/how to show that information from the default/original screen.

I also changed up the report screen; here are some WIP shots with faked data-

User just doing Archon's Forge or something and hasn't made any faction progress:
screen shot 2016-11-27 at 2 41 52 pm

User did some activities that gave faction rep:
screen shot 2016-11-27 at 2 42 36 pm

User ranked up a faction:
screen shot 2016-11-27 at 2 48 55 pm

I'm waiting on the badge stuff before committing again.

File: app/scripts/services/dimFarmingReportService.factory.js [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+
+  FarmingReportService.$inject = ['dimItemService', 'dimStoreService', '$interval'];
+
+    /**
+     * A service for tracking farming related things.
+     */
+  function FarmingReportService(dimItemService, dimStoreService, $interval) {
+    var intervalId;
+    var startTime;
+    var reportHashes = [
```
Comment (bhollis) [inline app/scripts/services/dimFarmingReportService.factory.js]: Having a hardcoded list like this makes me ☹️

File: app/scripts/services/dimFarmingReportService.factory.js [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+    }
+
+    return {
+      store: null,
+      start: function(store) {
+        var self = this;
+
+        self.store = store;
+
+        self.elapsed = "00:00:00";
```
Comment (bhollis) [inline app/scripts/services/dimFarmingReportService.factory.js]: We have a timer directive already that we can use.

File: app/scripts/services/dimFarmingReportService.factory.js [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+          var hours = Math.round(timeDiff % 24);
+
+          self.elapsed = pad(hours) + ":" + pad(minutes) + ":" + pad(seconds);
+        }, 1000);
+
+        // set up all the starting values
+        self.baseGlimmer = dimStoreService.getVault().glimmer;
+        self.baseMarks = dimStoreService.getVault().legendaryMarks;
+
+        self.baseVendors = {};
```
Comment (bhollis) [inline app/scripts/services/dimFarmingReportService.factory.js]: Let's not confuse vendors and character faction rep.

File: app/scripts/store/dimFarmItem.directive.js [hunk 4]
```diff
@@ -0,0 +1,56 @@
+(function() {
+  'use strict';
+
+  angular.module('dimApp')
+        .directive('dimFarmItem', dimItem);
```
Comment (bhollis) [inline app/scripts/store/dimFarmItem.directive.js]: I'd prefer if we didn't make yet another clone of `dimSimpleItem`. Can't we just set the amount and have simple item work for us?

File: app/scripts/store/dimFarmItem.directive.js [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+    template: `
+      <div class="unequipped" ng-class="{'farm-rank-up': vm.item.rankedUp}">
+        <div class="item" title="{{vm.item.faction.factionName}}\n{{vm.item.progressToNextLevel}}/{{vm.item.nextLevelAt}}\n{{'Level' | translate}}: {{vm.item.level}}">
+          <svg width="48" height="48">
+            <image xlink:href="" ng-attr-xlink:href="{{vm.item.faction.factionIcon | bungieIcon}}" width="48" height="48" />
+            <polygon stroke-dasharray="121.622368" ng-if="vm.item.progressToNextLevel > 0" style="stroke-dashoffset:{{121.622368-(121.622368*vm.item.progressToNextLevel/vm.item.nextLevelAt)}}" fill-opacity="0" stroke="#FFF" stroke-width="3" points="24,2.5 45.5,24 24,45.5 2.5,24" stroke-linecap="butt"/>
+          </svg>
+          <div class="item-stat item-faction">+{{vm.item.xpGain}}</div>
+        </div>
+      </div>`
```
Comment (bhollis) [inline app/scripts/store/dimFarmItem.directive.js]: Likewise if we're going to do this we should just make a faction-rep component that can be used multiple places.

File: app/scripts/services/dimFarmingReportService.factory.js (hunk 2 above)
Comment (Mulchman) [inline app/scripts/services/dimFarmingReportService.factory.js]: I'm not finding it. What's it called?

File: app/scripts/store/dimFarmItem.directive.js (hunk 5 above)
Comment (Mulchman) [inline app/scripts/store/dimFarmItem.directive.js]: There's some custom stuff I haven't checked in wrt this component. I've added a rank up background and click handler that ties in to `dimFarmingReportService`.

**Comment by Mulchman** [comment]
Here's how it's looking as of recent commits.

Summary-
farm-summary

* The engram icon glows when you hover over it. You can't see the cursor in the screenshot, but I'm hovering over the engram icon. The engram icon has become the badge/indicator. I had something else in mind but I don't know how to do it with HTML/CSS.

Report-
farm-report

* Used the back arrow like in other areas of the app.
* Reputation things get a background + border when you rank them up.
 * If you click a ranked up reputation thing it'll remove the background + border, like in the case you went to the Tower to get your package but didn't want to end farming mode.
* Messed with spacing of glimmer/legendary marks.

**Comment by Mulchman** [comment]
This is the idea I had for denoting a faction rank up ( @SunburnedGoose ) -> start adorning (up to 6?) the engram icon.

image

This is all faked out and manually positioned via Chrome styles box.

File: app/scripts/services/dimFarmingReportService.factory.js (hunk 2 above)
Comment (bhollis) [inline app/scripts/services/dimFarmingReportService.factory.js]: angular-timer

Comment (Mulchman) [inline app/scripts/services/dimFarmingReportService.factory.js]: Thanks!

**Comment by SunburnedGoose** [comment]
I like using the engram to toggle the summary screen.  The idea of moving factions around the icon when you rank up is interesting.  It would incentivze the user to click and discover the summary screen.

**Comment by kyleshay** [comment]
What if we made the icons on this page 1/4 of the size? I think an explicit button that toggles between the views would be nice. The dialog overall needs some work. Maybe it should be fixed somewhere else. Maybe it can expand/collapse like the rows of item categories? For the sake of moving this forward though I'm pretty happy with the changes. 

Does anyone else have any feedback?

**Comment by bhollis** [comment]
I feel like we should take a step back and try to come up with a good design for this feature, including what its goals are for the user, before going too much further. Do folks agree with that?

File: app/scripts/dimApp.i18n.js [hunk 6]
```diff
@@ -394,7 +399,12 @@
             },
             OutOfRoom: "Non hai più spazio per trasferire oggetti dal {{character}}. E' il momento di decriptare degli engrammi e pulire il cestino!",
             Quickmove: "Spostamento Rapido",
-            Stop: "Stop"
+            Stop: "Stop",
+            Report: {
```
Comment (delphiactual) [inline app/scripts/dimApp.i18n.js]: Do not add strings to foreign languages unless you are translating them.

**Comment by bhollis** [comment]
I chatted with @kyleshay and we decided to close this PR up for now. Thanks for the proposals, but it's a bit too much added complexity for now.

**Pull request closed without merging**